program-runtime/src/executor_cache.rs

```diff
@@ -14,80 +14,69 @@ use {
 ///
 /// The BankExecutorCache can not be updated directly as transaction batches are
 /// processed in parallel, which would cause a race condition.
 #[derive(Default, Debug)]
 pub struct TransactionExecutorCache {
-    pub executors: HashMap<Pubkey, TransactionExecutorCacheEntry>,
+    /// Executors which are visible to the transaction
+    pub executors: HashMap<Pubkey, Option<Arc<dyn Executor>>>,
+    /// Executors of programs which were deploymed during the transaction
+    pub deployments: HashMap<Pubkey, Arc<dyn Executor>>,
+    /// Executors which were loaded after the transaction was loaded (because they were missing in the cache)
+    pub loaded: HashMap<Pubkey, Arc<dyn Executor>>,
 }
 
 impl TransactionExecutorCache {
     pub fn new(executable_keys: impl Iterator<Item = (Pubkey, Arc<dyn Executor>)>) -> Self {
         Self {
             executors: executable_keys
-                .map(|(key, executor)| {
-                    let entry = TransactionExecutorCacheEntry {
-                        executor,
-                        difference: TxBankExecutorCacheDiff::None,
-                    };
-                    (key, entry)
-                })
+                .map(|(id, executor)| (id, Some(executor)))
                 .collect(),
+            deployments: HashMap::new(),
+            loaded: HashMap::new(),
         }
     }
 
-    pub fn get(&self, key: &Pubkey) -> Option<Arc<dyn Executor>> {
-        self.executors.get(key).map(|entry| entry.executor.clone())
+    pub fn get(&self, key: &Pubkey) -> Option<Option<Arc<dyn Executor>>> {
+        self.executors.get(key).cloned()
     }
 
-    pub fn set(&mut self, key: Pubkey, executor: Arc<dyn Executor>, replacement: bool) {
-        let difference = if replacement {
-            TxBankExecutorCacheDiff::Updated
+    pub fn set(
+        &mut self,
+        key: Pubkey,
+        executor: Arc<dyn Executor>,
+        upgrade: bool,
+        delay_visibility_of_program_deployment: bool,
+    ) {
+        if upgrade {
+            if !delay_visibility_of_program_deployment {
+                self.executors.insert(key, Some(executor.clone()));
+            } else {
+                // Do not overwrite an existing entry
+                self.executors.entry(key).or_insert(None);
+            }
+            self.deployments.insert(key, executor);
         } else {
-            TxBankExecutorCacheDiff::Inserted
-        };
-        let entry = TransactionExecutorCacheEntry {
-            executor,
-            difference,
-        };
-        let _was_replaced = self.executors.insert(key, entry).is_some();
+            self.executors.insert(key, Some(executor.clone()));
+            self.loaded.insert(key, executor);
+        }
     }
 
-    pub fn update_global_cache(
-        &self,
+    pub fn write_back_to_global_cache(
+        &mut self,
         global_cache: &RwLock<BankExecutorCache>,
-        selector: impl Fn(TxBankExecutorCacheDiff) -> bool,
+        deployments: bool,
```

Comment:
Overloading this function via a bool isn't very clear.  There isn't much overlap, how about making two functions, `write_back_deployments_to_global_cache` and `write_back_loaded_to_global_cache`.  Or maybe better yet a function that returns (via `into_iter()` either deployments or loaded and the caller calls `put()` on the global cache (rather then passing it in)?